### ISSUE TITLE
Enforce https in grin wallet

### ISSUE THREAD

**Issue by lehnberg** [mimblewimble/grin#2523]
I'd like to propose that https is enforced for transactions in grin wallet, and that http is deprecated due to it being insecure and exposing the transacting parties to unnecessarily easy attacks.

Related: https://github.com/vault713/wallet713/issues/89

**Comment by antiochp** [comment]
This may be a hard sell due to needing a cert for https?

**Comment by rsoltanzadeh** [comment]
How about giving a huge warning whenever it tries to use HTTP and making sure the users understand the risks every time (prompting once or twice unless `--no-warnings` or similar flag is used) rather than enforcing HTTPS? Even if certs were easy and effortless to get, CAs still have to be trusted. While that is normally not a problem, it might be in crypto. It might give a false sense of security to users as cryptocurrencies are otherwise trustless (or supposed to be).

**Comment by lehnberg** [comment]
If we do not set a minimum standard of what is acceptable security here (and then raise the bar to that level), no-one else will. And default instead becomes what's the easiest and least effort-taking to support, regardless of what that means to privacy and security.

I personally think http should never have been allowed in the first place on mainnet. This is not the type of transaction method we want to be the lowest common denominator in transaction standard between businesses and their users. It kind of defeats the entire purpose of why you would use Grin in the first place, in my opinion.

@rsoltanzadeh I don't think HTTPS is optimal by any means. But I'd rather the problem be that users _sometimes_ need to trust a certain CA for a cert's authenticity, rather than being at _constant_ risk of being trivially MITM'ed and/or have their transaction slates passed around the net unencrypted.

**Comment by mcdallas** [comment]
I know that http is bad but from what I can tell from discord the vast majority of users use this method for transacting. With https anyone that wants to receive coins will have to register a domain and get a Let's encrypt cert that needs to be updated every 90 days. Do we want to make it even *harder* to transact with grin?

Imagine most pools/exchanges not being able to send payouts anymore, they will most likely stop upgrading their nodes to not disrupt operations.

I think that until a more user friendly alternative comes, http needs to stay (maybe with a big fat warning).

**Comment by hashmap** [comment]
@mably It doesn't protect against mitm attack at all. Certificate is not an issue at all, the roadblock is domain name. However static ip is the biggest issue anyway for both http and https.

**Comment by mably** [comment]
Ok, it's harder to take over a DNS than an IP address, that's the idea?

**Comment by mably** [comment]
What if it could be implemented over Tor?

**Comment by hashmap** [comment]
No, self-signed certificate can't be checked against trusted authority, so if I intercept your traffic and sign it with my self-signed certificate your recipient can't tell the difference.
You can't get a free certificate for IP address atm and it's questionable in general, so it's cheaper/more secure to get a domain name. For dyn IP there are so many issues, most users can't even expose a port.
Tor should work, but it requires a manual intervention form a user and some technical skills.

**Comment by hashmap** [comment]
I think https server makes sense for business or advanced users (with static ip and domain name), https clients could be widely used to interact with such services. It's not hard to implement model where a "home user" only runs http client both for payment and withdrawals, perhaps we should standardize it and implement in the wallet.

**Comment by garyyu** [comment]
Related to https://github.com/mimblewimble/grin/issues/2224.

I would propose to have a transaction privacy enhancement solution, other than limit the transport methods.

**Comment by lehnberg** [comment]
@mcdallas:

> I know that http is bad but from what I can tell from discord the vast majority of users use this method for transacting.

Yes exactly. This is the problem - it risks becoming a major issue for us. It would be unfortunate if Grin becomes known as the currency where funds just get stolen out of thin air as you try to make deposits or withdrawals. It should never have been supported in the first place on mainnet.

> I think that until a more user friendly alternative comes, http needs to stay (maybe with a big fat warning).

It's also possible that as long as we allow this insecure yet comparatively convenient way of transferring, there will be no better alternatives. There will be little incentives to develop one or for users and exchanges to switch to it.

**Comment by kargakis** [comment]
> I think that until a more user friendly alternative comes, http needs to stay (maybe with a big fat warning).

+10000

We should be cautious about breaking people now that the ship has sailed. Also, the UX needs to get better, not worse.

**Comment by rsoltanzadeh** [comment]
What about including symmetric encryption in HTTP transfers?

>Alice: Bob, send me 5 grin. Code phrase: `"make me grin again"`
Bob: Alright, encrypted slate coming via HTTP right now!

**Comment by hashmap** [comment]
@rsoltanzadeh that's basically how TLS (HTTPS) works plus it includes keys exchange

**Comment by casey** [comment]
What about some form of public key + address scheme?

• Wallets generate a key pair and print out the hash of the public key
• Wallet URLs look like `PUBKEY-HASH@DOMAIN or `PUBKEY-HASH@IP-ADDRESS`
• Connections are authenticated and encrypted with the key pair, so MITM attacks are not possible.
• To avoid having to implement an entirely new authentication and encryption scheme, key pairs could actually be self-signed certificates. (There exists at least [one nice rust crate](https://crates.io/crates/rcgen) for generating self-signed certs.)

**Comment by rsoltanzadeh** [comment]
@hashmap the point was to not have to go through the hurdle of getting certificates, not that it's a better form of encryption. 

Include symmetric encryption on HTTP transfers and release it quickly and work on a HTTPS solution afterwards? This feels urgent.

**Comment by garyyu** [comment]
@casey I'm interested in your proposed `PUBKEY-HASH@DOMAIN/IP` solution. Would you like file a PR for it?

**Comment by kargakis** [comment]
> What about some form of public key + address scheme?

End users are still required to do port-forwarding in their routers and/or own domains with this approach, right? Creating an address-based scheme as a stopgap when TLS is already an option is likely going to distract both end-users and development effort. 

A NAT-friendly tx building solution is going to be more useful IMO: https://github.com/mimblewimble/grin/issues/1798

**Comment by casey** [comment]
@garyyu Unfortunately I have too much on my plate at the moment to submit a PR :(

@kargakis Users would still be required to do port-forwarding, but they wouldn't be required to own a domain or get a cert. I would very much like to see a NAT-friendly TX solution available, most likely tor or a tx relay, but this might be a good option in the mean time.

**Comment by ignopeverell** [comment]
_grumble_ _grumble_ embed i2pd _grumble_ grumble_

**Comment by kargakis** [comment]
@ignopeverell instead of embedding i2pd, could the wallet just use an i2p client? It would probably be simpler to implement and release at the expense of not contributing back to the i2p network.

**Comment by ignopeverell** [comment]
The additional advantage with embedding i2pd, which I think makes it worth it, is that you get a network address. So that would solve multiple issues in one fell swoop (not to mention the privacy gains).

**Comment by lehnberg** [comment]
Better late than never, this was [discussed during the Feb 5 dev meeting](https://github.com/mimblewimble/grin-pm/blob/master/notes/20190205-meeting-development.md#decision-approach-for-http-transactions-in-grin-wallet), and the following was agreed:  

> * Http command line send/receive to fail with a big warning and instruction to add `--unsafe` in the command.
> * Send `--unsafe -d http://..` executes, but also has a warning added.
> * Fully deprecate http once there's a sensible path to do so.
